internal/gengapic/client_init.go

```diff
@@ -135,155 +135,244 @@ func (g *generator) clientOptions(serv *descriptor.ServiceDescriptorProto, servN
 	p := g.printf
-
-	var hasLRO bool
+	p("// internal%sClient is an interface that defines the methods availaible from %s.", servName, g.apiName)
+	p("type internal%sClient interface {", servName)
 	for _, m := range serv.Method {
-		if g.isLRO(m) {
-			hasLRO = true
-			break
+		inType := g.descInfo.Type[m.GetInputType()]
+		inSpec, err := g.descInfo.ImportSpec(inType)
+		if err != nil {
+			return err
+		}
+		outType := g.descInfo.Type[m.GetOutputType()]
+		outSpec, err := g.descInfo.ImportSpec(outType)
+		if err != nil {
+			return err
 		}
-	}
 
-	imp, err := g.descInfo.ImportSpec(serv)
-	if err != nil {
-		return err
+		if m.GetOutputType() == emptyType {
+			p("%s(context.Context, *%s.%s, ...gax.CallOption) error",
+				m.GetName(),
+				inSpec.Name,
+				inType.GetName())
+			continue
+		}
+
+		if pf, err := g.pagingField(m); err != nil {
+			return err
+		} else if pf != nil {
+			iter, err := g.iterTypeOf(pf)
+			if err != nil {
+				return err
+			}
+			p("%s(context.Context, *%s.%s, ...gax.CallOption) *%s",
+				m.GetName(),
+				inSpec.Name,
+				inType.GetName(),
+				iter.iterTypeName)
+			continue
+		}
+
+		switch {
+		case g.isLRO(m):
+			// Unary call where the return type is a wrapper of
+			// longrunning.Operation and more precise types
+			p("%s(context.Context, *%s.%s, ...gax.CallOption) (*%s, error)",
+				m.GetName(), inSpec.Name, inType.GetName(), lroTypeName(m.GetName()))
+		case m.GetClientStreaming():
+			// Handles both client-streaming and bidi-streaming
+			p("%s(context.Context, ...gax.CallOption) (%s.%s_%sClient, error)",
+				m.GetName(), inSpec.Name, serv.GetName(), m.GetName())
+		case m.GetServerStreaming():
+			// Handles _just_ server streaming
+			p("%s(context.Context, *%s.%s, ...gax.CallOption) (%s.%s_%sClient, error)",
+				m.GetName(), inSpec.Name, inType.GetName(), inSpec.Name, serv.GetName(), m.GetName())
+		default:
+			// Regular, unary call
+			p("%s(context.Context, *%s.%s, ...gax.CallOption) (*%s.%s, error)",
+				m.GetName(), inSpec.Name, inType.GetName(), outSpec.Name, outType.GetName())
+		}
 	}
+	p("}")
+	p("")
+
+	p("// %s is a thin wrapper that holds an internal%sClient", servName, servName)
+	p("// It is agnostic as to the underlying transport, i.e. json+http, gRPC, or other.")
+	p("type %s struct {", servName)
+	p("    internal%sClient", servName)
+	p("}")
+	p("")
+
+	return nil
+}
+
+func (g *generator) grpcClientInit(serv *descriptor.ServiceDescriptorProto, servName string, imp pbinfo.ImportSpec, hasLRO bool) {
+	p := g.printf
 
 	// client struct
-	{
-		p("// %sClient is a client for interacting with %s.", servName, g.apiName)
-		p("//")
-		p("// Methods, except Close, may be called concurrently. However, fields must not be modified concurrently with method calls.")
-		p("type %sClient struct {", servName)
+	p("// %sClient is a client for interacting with %s over gRPC transport.", servName, g.apiName)
+	p("// It satisfies the %sAbstractClient interface.", servName)
```

Comment:
Probably don't want to document the private interface (and also there is no AbstractClient interface now?).

Reply:
Is not documenting internal structs/interfaces in the generated surface the norm? This is (mostly) to aid my own eyeball grepping when debugging generated clients.

Reply:
Well, for one, there is no `*AbstractClient` type defined anywhere, so I wouldn't be documenting that :)